tests/rewriter_test.py

```diff
@@ -1,265 +1,117 @@
-            [Constant.from_attr(IntegerAttr.from_int_and_width(43, 64), i32)])
+    def transformation(module: ModuleOp, rewriter: Rewriter) -> None:
+        constant_op = module.ops[0]
+        new_constant = Constant.from_int_constant(1, i32)
+        new_add = Addi.get(new_constant, new_constant)
 
-    @op_type_rewrite_pattern
-    def addi_rewrite(op: Addi,
-                     new_operands: List[SSAValue]) -> Optional[RewriteAction]:
-        return RewriteAction.from_op_list(
-            [Muli.get(op.input1.op, op.input2.op)])
+        rewriter.replace_op(constant_op, [new_constant, new_add])
 
-    rewrite_and_compare(
-        ctx, prog, expected,
-        PatternRewriteWalker(GreedyRewritePatternApplier([
-            AnonymousRewritePattern(constant_rewrite),
-            AnonymousRewritePattern(addi_rewrite)
-        ]),
-                             apply_recursively=False))
+    rewrite_and_compare(prog, expected, transformation)
 
 
-def test_operation_deletion():
-    """Test rewrites where SSA values are deleted."""
-    ctx = MLContext()
-    builtin = Builtin(ctx)
-    std = Std(ctx)
-    arith = Arith(ctx)
-
+# Test an operation replacement with manually specified results
+def test_replace_op_new_results():
     prog = \
-"""module() {
-  %0 : !i32 = arith.constant() ["value" = 5 : !i32]
+    """module() {
+%0 : !i32 = arith.constant() ["value" = 2 : !i32]
+%1 : !i32 = arith.addi(%0 : !i32, %0 : !i32)
 }"""
 
     expected = \
-"""module() {}"""
-
-    @op_type_rewrite_pattern
-    def match_and_rewrite(
-            op: Constant,
-            new_operands: Optional[List[SSAValue]]) -> Optional[RewriteAction]:
-        return RewriteAction([], [None])
-
-    rewrite_and_compare(
-        ctx, prog, expected,
-        PatternRewriteWalker(AnonymousRewritePattern(match_and_rewrite)))
-
-
-def test_operation_deletion_failure():
-    """Test rewrites where SSA values are deleted."""
-    ctx = MLContext()
-    builtin = Builtin(ctx)
-    std = Std(ctx)
-    arith = Arith(ctx)
-
-    prog = \
 """module() {
-  %0 : !i32 = arith.constant() ["value" = 5 : !i32]
+  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
   %1 : !i32 = arith.addi(%0 : !i32, %0 : !i32)
+  %2 : !i32 = arith.addi(%1 : !i32, %1 : !i32)
 }"""
 
-    @op_type_rewrite_pattern
-    def match_and_rewrite(
-            op: Constant,
-            new_operands: Optional[List[SSAValue]]) -> Optional[RewriteAction]:
-        return RewriteAction([], [None])
+    def transformation(module: ModuleOp, rewriter: Rewriter) -> None:
+        constant_op = module.ops[0]
+        new_constant = Constant.from_int_constant(1, i32)
+        new_add = Addi.get(new_constant, new_constant)
 
-    parser = Parser(ctx, prog)
-    module = parser.parse_op()
-    walker = PatternRewriteWalker(AnonymousRewritePattern(match_and_rewrite))
+        rewriter.replace_op(constant_op, [new_constant, new_add],
+                            [new_add.output])
```

Comment:
Maybe change this test such that adding the additional argument actually matters, e.g. replace `new_add.output`with `new_constant.output`, such that the two adds basically become the same.

Reply:
Good point!